### PR TITLE
Better arviz support in base inference

### PR DESCRIPTION
**Pull request by zaxtax**
### Motivation
When I was doing bug fixing I noticed my arviz changes were lost in the shuffle. These commits add the functionality while keeping the APIs unchanged as much as possible.

### Changes proposed
Changes the

### Test Plan
Please provide clear instructions on how the changes were verified. Attach screenshots if applicable.

### Types of changes
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **[CONTRIBUTING](https://github.com/facebookresearch/beanmachine/blob/main/CONTRIBUTING.md)** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

**Comment by horizon-blue** [comment]
:D Ah, this is a nice feature to have for our next release. Thanks for working on this, Rob! I'm going to import the PR now just so that we can see the test results. I'll have to review this a bit later, though, since there're a few other pre-launch items that we have to focus on at the moment.

In the mean time, there's a `get_chain` method that returns another `MonteCarloSamples` that people sometimes use to retrieve the data for a particular chain. Could you check if that's still working as expected with your change?

**Comment by zaxtax** [comment]
I think `get_chain` should still work like it did before, but I should pass likelihoods to the `MonteCarloSamples` object returned by that method.

**Comment by horizon-blue** [comment]
@zaxtax Was facebook-github-bot acting crazy? Or did you actually update this PR 42 times this afternoon? 🤣

**Comment by zaxtax** [comment]
I force-pushed once and I think I upset the bot. I apologise to everyone
else you got that flood of emails 😅

**Comment by horizon-blue** [comment]
Haha don't worry, this is actually hilarious. I'm going to report this bug to our internal support team :)

File: src/beanmachine/ppl/inference/monte_carlo_samples.py [hunk 1]
```diff
@@ -77,7 +95,21 @@ def get_chain(self, chain: int = 0) -> "MonteCarloSamples":
...
+        else:
+            logll = {
+                rv: self.get_log_likelihoods(rv, True)[[chain]]
+                for rv in self.log_likelihoods
+            }
+
+        new_mcs = MonteCarloSamples(
+            chain_results=samples,
+            num_adaptive_samples=self.num_adaptive_samples,
+            logll_results=logll,
```
Comment (horizon-blue) [inline src/beanmachine/ppl/inference/monte_carlo_samples.py:110]: Our internal linter (Pyre) is complaining about the following typing issue:
> Incompatible parameter type [6]: Expected `Optional[List[Dict[RVIdentifier, torch.Tensor]]]` for 3rd parameter `logll_results` to call `MonteCarloSamples.__init__` but got `Optional[Dict[typing.Any, typing.Any]]`.

Speaking of which, do you mind adding a call to `get_chain` to the unit test as well? Thanks :D

Comment (zaxtax) [reply src/beanmachine/ppl/inference/monte_carlo_samples.py:110]: That should fix the linter error.

**Comment by horizon-blue** [comment]
I thought we still have a type error on this diff but seems like I was wrong :). Our internal tool does warn us that this PR is a bit old, so I'm going to rebase and rerun the internal tests. If the signal is still green then we are good to go. Thanks again for working on this, Rob!

**Pull request closed without merging**